test/e2e/upgrade/alert/alert.go

```diff
@@ -0,0 +1,167 @@
+}
+
+// Setup creates parameters to query Prometheus
+func (t *UpgradeTest) Setup(f *framework.Framework) {
+	g.By("Setting up post-upgrade alert test")
+
+	// Must use version that's can run within Ginkgo It
+	oc := exutil.NewCLIWithoutNamespaceInside("prometheus")
+
+	var ok bool
+	url, bearerToken, ok := helper.LocatePrometheus(oc)
+	if !ok {
+		framework.Failf("Prometheus could not be located on this cluster, failing test %s", t.Name())
+	}
+	t.url = url
+	t.bearerToken = bearerToken
+	t.oc = oc
+	framework.Logf("Post-upgrade alert test setup complete")
+}
+
+// Test checks if any critical alerts are firing.
+func (t *UpgradeTest) Test(f *framework.Framework, done <-chan struct{}, upgrade upgrades.UpgradeType) {
+	g.By("Checking for critical alerts")
+
+	// Recover current test if it fails so test suite can complete
+	defer g.GinkgoRecover()
+
+	// Block until upgrade is done
+	g.By("Waiting for upgrade to finish before checking for critical alerts")
+	<-done
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// Additonal delay after upgrade completion
+	g.By("Waiting before checking for critical alerts")
+	time.Sleep(alertCheckSleep)
+	cancel()
+
+	if helper.TestUnsupportedAllowVersionSkew() {
+		e2eskipper.Skipf("Test is disabled to allow cluster components to have different versions, and skewed versions trigger multiple other alerts")
+	}
+	t.oc.SetupProject()
+	ns := t.oc.Namespace()
+	execPod := exutil.CreateCentosExecPodOrFail(t.oc.AdminKubeClient(), ns, "execpod", nil)
+	defer func() {
+		t.oc.AdminKubeClient().CoreV1().Pods(ns).Delete(ctx, execPod.Name, *metav1.NewDeleteOptions(1))
+	}()
+
+	var errors map[string]error
+
+	// First check for any critical severity alerts that has occurred within the last alertPeriodCheckMinutes
+	query := fmt.Sprintf(`count_over_time(ALERTS{alertname!~"Watchdog|AlertmanagerReceiversNotConfigured|KubeAPILatencyHigh",alertstate="firing",severity="critical"}[%dm]) >= 1`, alertPeriodCheckMinutes)
+	errors = runQuery(f, query, false, t.oc, ns, execPod.Name, t.url, t.bearerToken)
+
+	// If there were no critical alerts ensure Prometheus has been up and running for entire post-upgrade
+	// period by verifying Watchdog alert has been in firing state
+	if len(errors) == 0 {
+		query := fmt.Sprintf(`count_over_time(ALERTS{alertstate="firing",alertname="Watchdog", severity="none"}[%dm]) : true`, alertCheckSleepMinutes)
+		errors = runQuery(f, query, true, t.oc, ns, execPod.Name, t.url, t.bearerToken)
+	}
+	if len(errors) != 0 {
+		exutil.DumpPodLogsStartingWith("prometheus-0", t.oc)
+	}
+	o.Expect(errors).To(o.BeEmpty())
+}
+
+// Teardown cleans up any remaining resources.
+func (t *UpgradeTest) Teardown(f *framework.Framework) {
+	// rely on the namespace deletion to clean up everything
+}
+
+func runQuery(f *framework.Framework, query string, expected bool, oc *exutil.CLI, ns, execPodName, baseURL, bearerToken string) map[string]error {
```

Comment:
This method is super common, I would go find one of these and move it into a separate package that everyone shares